### PR TITLE
Introduce more domain concepts

### PR DESCRIPTION
**Pull request by philandstuff**
Having introduced the register abstraction in #48, this continues that work by introducing more abstractions for pieces of our domain:

 - EntryLog, which encapsulates the append-only list of entries and the verifiable log of proofs
 - ItemStore, which encapsulates the store of items addressed by hash
 - RecordIndex, which encapsulates the linking of primary keys to current entries

They are still somewhat interrelated through transactions: you shouldn't be able to add a dangling entry or item so we often need to pass the same Handle into the ItemStore and the EntryLog to ensure they are both updated together, atomically.

As @johnollier suggested on #48, the name "PostgresRegister" is increasingly inaccurate at this stage but I can't think of a better name.

### CODE REVIEW

File: src/main/java/uk/gov/register/core/PostgresRegister.java [hunk 1]
```diff
@@ -20,107 +20,113 @@
...
     @Override
     public List<Record> max100RecordsFacetedByKeyValue(String key, String value) {
-        return registerDAO.findMax100RecordsByKeyValue(key, value);
+        return dbi.withHandle(h -> recordIndex.findMax100RecordsByKeyValue(h, key, value));
     }
 
     @Override
     public RegisterProof getRegisterProof() throws NoSuchAlgorithmException {
-        return verifiableLogService.getRegisterProof();
+        return dbi.withHandle(entryLog::getRegisterProof);
```
Comment (karlbaker02) [inline src/main/java/uk/gov/register/core/PostgresRegister.java]: We'll want to double check that this is wrapped in a transaction, or explicitly open a transaction

File: src/main/java/uk/gov/register/core/PostgresRegister.java [hunk 2]
```diff
@@ -20,107 +20,113 @@
...
     }
 
     @Override
     public EntryProof getEntryProof(int entryNumber, int totalEntries) {
-        return verifiableLogService.getEntryProof(entryNumber, totalEntries);
+        return dbi.withHandle(handle ->
```
Comment (karlbaker02) [inline src/main/java/uk/gov/register/core/PostgresRegister.java]: We'll want to double check that this is wrapped in a transaction, or explicitly open a transaction

File: src/main/java/uk/gov/register/core/PostgresRegister.java [hunk 3]
```diff
@@ -20,107 +20,113 @@
...
+                entryLog.getEntryProof(handle, entryNumber, totalEntries));
     }
 
     @Override
     public ConsistencyProof getConsistencyProof(int totalEntries1, int totalEntries2) {
-        return verifiableLogService.getConsistencyProof(totalEntries1, totalEntries2);
+        return dbi.withHandle(handle ->
```
Comment (karlbaker02) [inline src/main/java/uk/gov/register/core/PostgresRegister.java]: We'll want to double check that this is wrapped in a transaction, or explicitly open a transaction

**Comment by philandstuff** [comment]
rebased.

File: src/main/java/uk/gov/register/RegisterApplication.java [hunk 4]
```diff
@@ -101,15 +95,17 @@ public void run(RegisterConfiguration configuration, Environment environment) th
             @Override
             protected void configure() {
                 bind(entryQueryDAO).to(EntryQueryDAO.class);
```
Comment (michaelabenyohai) [inline src/main/java/uk/gov/register/RegisterApplication.java]: Can we remove this line now?

Comment (philandstuff) [inline src/main/java/uk/gov/register/RegisterApplication.java]: yes! opened #54 for this.

**Pull request merged**